crates/compiler/mono/src/drop_specialization.rs

```diff
@@ -1090,114 +1088,125 @@ fn specialize_list<'a, 'i>(
-                            // Should only happen when list elements are discarded.
-                            None => {
-                                let field_symbol =
-                                    environment.create_symbol(ident_ids, &format!("field_val_{i}"));
-
-                                let index_symbol =
-                                    environment.create_symbol(ident_ids, &format!("index_val_{i}"));
-
-                                let dec = arena.alloc(Stmt::Refcounting(
-                                    ModifyRc::Dec(field_symbol),
-                                    newer_continuation,
-                                ));
-
-                                let index = arena.alloc(Stmt::Let(
-                                    field_symbol,
-                                    Expr::Call(Call {
-                                        call_type: CallType::LowLevel {
-                                            op: LowLevel::ListGetUnsafe,
-                                            update_mode: UpdateModeId::BACKEND_DUMMY,
-                                        },
-                                        arguments: arena.alloc([*symbol, index_symbol]),
-                                    }),
-                                    item_layout,
-                                    dec,
-                                ));
-
-                                newer_continuation = arena.alloc(Stmt::Let(
-                                    index_symbol,
-                                    Expr::Literal(Literal::Int(i128::to_ne_bytes(i as i128))),
-                                    Layout::isize(layout_interner.target()),
-                                    index,
-                                ));
-                            }
-                        };
-                    }
-
-                    newer_continuation
-                }
-                _ => keep_original_decrement!(),
-            }
-        }
-        _ => {
-            // List length is unknown or the children are not reference counted, so we can't specialize.
-            keep_original_decrement!()
-        }
-    }
+    // TODO: Investigate if there is still any utility in List drop specialization.
+    // Fundamentally, an allocation for a list holds onto exactly one reference to all elements in the allocation.
+    // So even if there are 100 references to the list, the list will hold only 1 reference to each element.
+    // On top of that, the list now holds onto a reference to dead elements.
+    // A single slice to 10 elements of a 100 element list will hold 1 reference to each to the 100 elements.
+    // The whole allocation will get freed in a unit.
+    // This makes it much closer to the lifetime of the underlying vector for a reference slice in rust.
+    // If this specialization does not make sense anymore delete it as a whole.
```

Comment:
Anyone have thoughts on this? My gut feeling is that List just don't make sense to drop specialize anymore. That said, if we know a list is truly referencing all elements (no dead elements waiting to be freed), this optimization is technically correct still. Or I guess it could be applied if we knew the number of dead elements as well.